---
title: "MS_analysis_MQ_summary"
author: "Sam Siljee"
date: '2022-08-22'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Packages
```{r packages, include=FALSE}
library("tidyverse")
```

## Importing raw data
```{r reading in MQ results from the combined folder, include=FALSE}
# importing evidence file, this is the file suggested as the main analysis file
evidence <- read_tsv("~/Coding/MS_analysis/data/combined/txt/evidence.txt") %>%
  as.data.frame()
# removing decoys
evidence <- evidence[-which(evidence$Reverse == "+"),]
# adding column with first protein name if multiple protein matches
evidence$firstProtName <- sapply(str_split(evidence$`Protein names`,";",),'[',1)
# doing the same for multiple gene name matches
evidence$firstGeneName <- sapply(str_split(evidence$`Gene names`,";",),'[',1)
# converting the "Potential contaminant" column to a vector. The "Potential contaminant" column is unhelpfully laid out, with contaminants labeled as "+". and non-contaminants labeled as NA. Hence the slightly awkward phrasing of the below code, which takes the inverse of logical return from `is.na`
evidence$potCon <- !is.na(evidence$`Potential contaminant`)

# importing all peptides
allPeptides <- read_tsv("~/Coding/MS_analysis/data/combined/txt/allPeptides.txt") %>%
  as.data.frame()

# importing phospho sites
phosphoSites <- read_tsv("~/Coding/MS_analysis/data/combined/txt/Phospho (STY)Sites.txt") %>%
  as.data.frame()
# removing decoys
phosphoSites <- phosphoSites[-which(phosphoSites$Reverse == "+"),]
# adding column with first protein name if peptide matches multiple proteins
phosphoSites$firstProtName <- sapply(str_split(phosphoSites$`Protein names`,";",),'[',1)

# extracting basal cell unique markers list from "The Human Airway Epithelial Basal Cell Transcriptome, Hacket et al. PLoS ONE, 2011). Adding p63 and Keratin 5 manually
basalCellGenes <- read.csv("~/Coding/MS_analysis/tracked_files/basalCellTranscriptome.csv") %>%
  .$Gene.symbol %>% 
  c(., "TP63", "KRT5") %>% 
  sort()
```

## Enrichment for phosphopeptides
What proportion of results are enriched for phospho groups

```{r Phospho enrichment, echo=FALSE}
print("Fraction of results with at least one phospho modification") 
mean(evidence$`Phospho (STY)` > 0)
```

# Identifying basal cell markers
Listing markers found in phosphoproteome, along with fraction of markers found
```{r identifying basal cell markers, echo=FALSE}
markersPresent <- evidence %>%
  filter(firstGeneName %in% basalCellGenes) %>%
  .$firstProtName %>%
  unique() %>%
  sort()
print("Basal cell markers found")
markersPresent

print("Proportion of basal cell markers found")
length(markersPresent)/length(basalCellGenes)
```

## Contaminants

# Looking at fraction of contaminants
Dividing the entries in contaminants marked as "potential contaminant" by total number of entries

```{r Summarising fraction of contaminants, echo=FALSE}
print("Fraction of results categorised as contaminants")
length(which(evidence$`Potential contaminant` == "+")) / nrow(evidence)
```

# Listing contaminants
Listing all the potential contaminants from the phospho sites data, note that at least Keratin 5 is found in basal cells, and is therefore expected

```{r looking at the identified contaminants, echo=FALSE}
#contaminants in the phosphoSites data, first creating a string of all the protein names
print("List of contaminants found")
evidence %>%
  filter(potCon == TRUE) %>%
  .$firstProtName %>%
  unique() %>%
  sort()
```
